### PR TITLE
EFI workflow

### PR DESCRIPTION
**Pull request by helge22a**
## Description

This branch includes 5 scripts that will run the EFI forecast start to finish. This includes two met downloading scripts (EFI_dataprep and EFI_metprocess) one workflow script (EFI_workflow) this includes output processing and two function scripts (half_hour_downscale and efi_data_process).
## Motivation and Context

The idea is to make to easier for individuals to run the EFI workflow for the Forecast Challenge.
## Review Time Estimate

- [ ] Immediately
- [ ] Within one week
- [x] When possible
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue) 
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: scripts/EFI_workflow.R
Comment (mdietze) [inline scripts/EFI_workflow.R:203]: reference functions by namespace rather than using `library` loads

File: scripts/EFI_workflow.R
Comment (mdietze) [inline scripts/EFI_workflow.R:199]: 1. the goal of this remaining section of code will be unclear to most users
2. if the goal is to take output files that are in PEcAn standard and convert them to EFI's ensemble CSV standard, I'd make that into a function and put it into an appropriate package (utils?)
3. what's the difference between the PEcAn standard and the EFI (preferred) netcdf standard? in other words, is there even a need for this code? And if there is, why not use the EFI netcdf standard instead of the csv standard? Maybe there should be an issue (and discussion on the next PEcAn call) about reconciling the differences between the PEcAn and EFI standards so that we can avoid this step altogether.

File: scripts/EFI_workflow.R
Comment (mdietze) [inline scripts/EFI_workflow.R:8]: where possible, reference functions by namespace rather than loading libraries

**Review by mdietze — CHANGES_REQUESTED** [review]

File: scripts/EFI_dataprep.R
Comment (helge22a) [reply scripts/EFI_dataprep.R:117]: So index is not used anywhere else, my guess is that this script was supposed to be run immediately before the EFI workflow and that the index object was used there (since it would be already loaded into the environment) but since we changed the workflow I don't think we need this chunk anymore.

Comment (helge22a) [reply scripts/EFI_dataprep.R:117]: This chunk is still part of the for loop so let me try running the loop without it and see if that creates any issues

File: scripts/EFI_metprocess.R
Comment (helge22a) [reply scripts/EFI_metprocess.R:9]: Right so the I think we should keep the sourced files for now, only because this was one of the ways I got around the error with download.NOAA_GEFS.R. Even though the function the set to @export  if you don't source it the convert.input section of met.process cannot find the function. This seems like an issue in it of itself and something I'm working on. But for now sourcing the scripts is the only way to get this met.process.R script to run.

File: scripts/EFI_dataprep.R
Comment (helge22a) [reply scripts/EFI_dataprep.R:12]: Yes so I have added a comment in the script explaining what is in this .csv file. It is the sitename, site id, etc for whichever sites you want to run through the script. In this case it is the 4 EFI sites plus HARV. I can add the .csv I am using for the script.

File: scripts/EFI_metprocess.R
Comment (helge22a) [reply scripts/EFI_metprocess.R:15]: That's a great idea I will adjust the .csv file accordingly

File: scripts/EFI_workflow.R
Comment (helge22a) [reply scripts/EFI_workflow.R:24]: So my only hesitation with that and the reason I let it hard coded, is because right now between the debugging and catching up on running forecasts now that the met data collection scripts are working it was easier to have it hard coded so I could change the date to fit the situation. 

Also we have some sites we want to run daily and other monthly so that was a another reason I left the dates as hard coded. But I think now that the met download scripts are working I want to transition the dates to args, but given the variety of dates you might want to run with this workflow I am unsure the best way to have them as args?

**Review by mdietze — CHANGES_REQUESTED** [review]
you also need to do a `make document` after you add the missing exports

File: scripts/EFI_workflow.R
Comment (mdietze) [reply scripts/EFI_workflow.R]: 1. we'd agreed that we were running everything daily
2. this script is unusable in a cron job until you remove hard-coded dates. It was fine to hard-code during debugging, but now that the code is running you need to clean stuff like this up
3. If you pass the dates in as external args, then you can run the script for any dates.

File: scripts/EFI_workflow.R
Comment (helge22a) [reply scripts/EFI_workflow.R:95]: Yes I need to do this.

File: scripts/EFI_workflow.R
Comment (helge22a) [reply scripts/EFI_workflow.R:205]: Got it.

File: scripts/EFI_workflow.R
Comment (helge22a) [reply scripts/EFI_workflow.R:201]: So the primary reason I put this at the end of the workflow is so that if someone is running the workflow and wants to use the results of the model runs for statistical analysis or graphs you end up with a usable .csv. And so they do not need another script to do that. This is convenient for the EFI forecasts of course because it will already be in the EFI format but I thought it would be helpful in a broader sense too?

Comment (mdietze) [reply scripts/EFI_workflow.R:201]: I'm not saying that it's not useful, I'm saying you could make it more useful by (A) making it a function (B) documenting what that function aims to do and (C) use the preferred EFI standard (netCDF) rather than the tier 2 csv. The last change can come in a future PR

Comment (helge22a) [reply scripts/EFI_workflow.R:204]: Got it

**Comment by mdietze** [comment]
Remaining issues with PR:
* in EFI_dataprep.R, possible stray } at end
* EFI_workflow.R: still need to pass in site XML and start/end date info as an argument, rather than hard coding
Things to do for a future PR:
* Move more code into functions
* Better document scripts (e.g. efi_data_process.R)
* remove extraneous source calls
* remove library calls and call functions by namespace

**Review by mdietze — APPROVED** [review]

**Review by infotroph — CHANGES_REQUESTED** [review]
I wonder if it might be better to create a new directory for some (all?) of these files where the whole workflow can be kept together, rather than trying to fits the pieces into the existing PEcAn directory structure. Points that especially jumped out at me:

* The `settings` package provides R functions for *manipulating* PEcAn settings, so `base/settings/examples/` is for files that show how the settings functions work and is definitely not the right place to *store* large numbers of settings files. At the moment our largest collection of settings files is the `tests/` directory, but these seem different enough to deserve their own home.
* The `scripts/` directory is mostly used for administrative tools used during installation/updating of the PEcAn system as a whole, rather than for scripts that run particular workflows. But I did say "mostly" and it already contains five `workflow.*.R, so maybe there's precedent for adding the EFI workflows here.
* Whether or not the workflow scripts belong in `scripts/`, it would be pretty confusing for `dataprep_10_sites.csv` to live there as a loose dataset -- after this is merged, someone looking at the contents of the folder for the first time would need to search through all the scripts in the directory to figure out what it's used for. This is what tips the balance for me to suggest a whole new EFI folder where scripts, settings files, and data can all live together. 

I'm marking this as a requested change, but will defer to @mdietze if he feels strongly that these are OK to merge at their current paths.

File: modules/data.atmosphere/R/download.raw.met.module.R
Comment (infotroph) [inline modules/data.atmosphere/R/download.raw.met.module.R:28]: I usually expect function names starting with dots or underscores to be for internal use only. Are we sure we want to export this?

File: modules/data.atmosphere/R/met.process.R
Comment (infotroph) [reply modules/data.atmosphere/R/met.process.R:15]: On a close look, more important than the duplicated export is that the description of the `overwrite` parameter is supposed to continue for 8 more lines (all the way to the `@author` tag), and adding `@export` here interrupts that and causes the man page to be missing some crucial information.

**Comment by infotroph** [comment]
To contradict @mdietze again 😅:  How sure are we that the helper functions need to be exported? If I searched right, none of them are used in the scripts in this PR, so unless they're definitely needed for a planned next step then I'd suggest *not* exporting them now.

As part of some ongoing work on reducing dependencies between packages, I've been advocating for a principle that functions are only exported when we're sure they need to be used outside the package they're defined in. If these are "helper functions" in the sense that they're provided for use by other GEFS functions that live inside data.atmosphere, they probably do not need to be exported now and are easy enough to export later when it's needed.

**Comment by mdietze** [comment]
@infotroph yeah, we probably exported more than we strictly had to, but in practice the met.process workflow wasn't running outside of do.conversions unless @helge22a `source`ed these functions, so I had her export the functions instead.

**Comment by infotroph** [comment]
>  in practice the met.process workflow wasn't running outside of do.conversions unless @helge22a sourceed these functions

@mdietze That's very weird! But I still see `source()` calls in these scripts -- are you sure the exporting was actually what helped?

**Comment by infotroph** [comment]
Also: Is there a difference between the settings files inside and outside of the `Site_XMLs/` directory, e.g. `base/settings/examples/Site_XMLS/harvard.xml` vs `base/settings/examples/harvard.xml`?

**Comment by mdietze** [comment]
@infotroph yes, @helge22a still needs to clean up those calls to `source`. FYI this is a first PR from a new student.

File: modules/data.atmosphere/R/met.process.R
Comment (helge22a) [reply modules/data.atmosphere/R/met.process.R]: I changed the location of the @export

File: modules/data.atmosphere/R/download.raw.met.module.R
Comment (helge22a) [reply modules/data.atmosphere/R/download.raw.met.module.R:28]: Adding the @export to this function makes it easier to run met.process outside of do_conversions.

**Review by infotroph — APPROVED** [review]

**Pull request merged**
